### ISSUE TITLE
kubernetes_resource_quota variable interpolation does not work with terraform v0.12.0

### ISSUE THREAD

**Issue by conet** [terraform-providers/terraform-provider-kubernetes#452]
### Terraform Version
```
terraform -v
Terraform v0.12.0
+ provider.google v2.7.0
+ provider.google-beta v2.7.0
+ provider.kubernetes v1.7.0
+ provider.template v2.1.2
```

### Affected Resource(s)
- kubernetes_resource_quota

### Expected Behavior
This should work:
```
resource "kubernetes_resource_quota" "resource" {
  metadata {
    name = "${var.namespace}-quota"
    namespace = "${var.namespace}"
  }

  spec {
    hard = {
      "pods"            = "${var.quota_pods}"
      "requests.cpu"    = "${var.quota_cpu_request}"
      "requests.memory" = "${var.quota_memory_request}"
      "limits.cpu"      = "${var.quota_cpu_limit}"
      "limits.memory"   = "${var.quota_memory_limit}"
    }
  }
}
```

### Actual Behavior
I'm getting this error:
```
Error: spec.0.hard.pods ("74D93920-ED26-11E3-AC10-0800200C9A66"): quantities must match the regular expression '^([+-]?[0-9.]+)([eEinumkKMGTP]*[-+]?[0-9]*)$'

  on ../../modules/module-name/main.tf line 11, in resource "kubernetes_resource_quota" "resource":
  11: resource "kubernetes_resource_quota" "resource" {
```

If I place the values in line it works:
```
resource "kubernetes_resource_quota" "resource" {
  metadata {
    name = "${var.namespace}-quota"
    namespace = "${var.namespace}"
  }

  spec {
    hard = {
      "pods"            = 5    #"${var.quota_pods}"
      "requests.cpu"    = "3"  #"${var.quota_cpu_request}"
      "requests.memory" = "4Gi"#"${var.quota_memory_request}"
      "limits.cpu"      = "4"  #"${var.quota_cpu_limit}"
      "limits.memory"   = "8Gi"#"${var.quota_memory_limit}"
    }
  }
}
```

It seems like the variable reference is tested agains the regexp not it's value.

**Comment by alexsomesan** [comment]
Try without the "${..}" interpolation syntax. In Terraform 0.12, variables and resource attributes don't need to be wrapped like that anymore when passed as values.

**Comment by alexsomesan** [comment]
Here is an explanation of what I just mentioned: https://www.hashicorp.com/blog/terraform-0-12-preview-first-class-expressions

**Comment by conet** [comment]
I already did that, the results are the same.

**Comment by alexsomesan** [comment]
Can you also paste the HCL fragment that sets those variables?

**Comment by conet** [comment]
```
variable "quota_pods" {
  default = "5"
}

variable "quota_memory_limit" {
  default = "8Gi"
}

variable "quota_memory_request" {
  default = "4Gi"
}

variable "quota_cpu_limit" {
  default = "4"
}

variable "quota_cpu_request" {
  default = "3"
}
```
I'm currently using default values.

**Comment by alexsomesan** [comment]
On first look the regular expression used for validation seems wrong.

Can you try expressing memory as 4G instead of 4Gi.

I will validate that as soon as I get off my flight.

**Comment by conet** [comment]
I don't think that the values are wrong, did you see this remark?

> If I place the values in line it works

**Comment by conet** [comment]
Just to rule out any issues I tried only setting the `pod` attribute, inline it works if I use a variable it fails with that error.

**Comment by alexsomesan** [comment]
I've looked a bit further into this and it seems to point to an issue in Terraform core related to the handling of validation functions.
I've documented the behaviour in the issue linked above and it will be looked into by the core engineers.

**Comment by conet** [comment]
Thanks for reporting it, I didn't know that it was something related to the core, I only noticed it when using this resource, probably because the way this resource is defined.

**Comment by pdecat** [comment]
Does it work with unquoting the variable references?

E.g.:

```
resource "kubernetes_resource_quota" "resource" {
  metadata {
    name = "${var.namespace}-quota"
    namespace = "${var.namespace}"
  }

  spec {
    hard = {
      "pods"            = var.quota_pods
      "requests.cpu"    = var.quota_cpu_request
      "requests.memory" = var.quota_memory_request
      "limits.cpu"      = var.quota_cpu_limit
      "limits.memory"   = var.quota_memory_limit
    }
  }
}
```

**Comment by drzero42** [comment]
I am having the same issue when setting the size for a `kubernetes_persistent_volume`. If I use a var it gives the error. If I use a literal string, it works.

This:
```
variable "disk_size" {
  type        = "string"
  default     = "200Gi"
}

resource "kubernetes_persistent_volume" "disk" {
...
  spec {
    capacity = {
      storage = var.disk_size
    }
...
```
Gives this error: `Error: spec.0.capacity.storage ("74D93920-ED26-11E3-AC10-0800200C9A66"): quantities must match the regular expression '^([+-]?[0-9.]+)([eEinumkKMGTP]*[-+]?[0-9]*)$'`

Whereas this:
```
resource "kubernetes_persistent_volume" "disk" {
...
  spec {
    capacity = {
      storage = "200Gi"
    }
...
```

Works fine.

**Comment by conet** [comment]
> Does it work with unquoting the variable references?

@pdecat no, see [this comment](https://github.com/terraform-providers/terraform-provider-kubernetes/issues/452#issuecomment-495302329)

**Comment by pdecat** [comment]
Sorry, I missed that.

Looks like the issue will be [fixed in terraform v0.12.1](https://github.com/hashicorp/terraform/issues/21419).

**Comment by yob** [comment]
I've run into this issue with the following resource:

```hcl
variable "quota_cpu" {
  default = "4"
}

variable "quota_memory" {
  default = "10Gi"
}

resource "kubernetes_resource_quota" "limit-quota" {
  metadata {
    name      = "limit-quota"
    namespace = kubernetes_namespace.app-namespace.metadata[0].name
  }

  spec {
    hard = {
      "limits.memory" = var.quota_memory
      "limits.cpu"    = var.quota_cpu
    }
  }
}
```

```bash
$ terraform-0-12-1 plan

Error: spec.0.hard.limits.cpu ("74D93920-ED26-11E3-AC10-0800200C9A66"): quantities must match the regular expression '^([+-]?[0-9.]+)([eEinumkKMGTP]*[-+]?[0-9]*)$'                                                                                                                        

  on .terraform/modules/cluster-services/main.tf line 153, in resource "kubernetes_resource_quota" "limit-quota":
 153: resource "kubernetes_resource_quota" "limit-quota" {



Error: spec.0.hard.limits.memory ("74D93920-ED26-11E3-AC10-0800200C9A66"): quantities must match the regular expression '^([+-]?[0-9.]+)([eEinumkKMGTP]*[-+]?[0-9]*)$'                                                                                                                     

  on .terraform/modules/cluster-services/main.tf line 153, in resource "kubernetes_resource_quota" "limit-quota":
 153: resource "kubernetes_resource_quota" "limit-quota" 
```

I'm running terraform 0.12.1 - is it possible the fix isn't complete?

**Comment by pdecat** [comment]
Looks like I was wrong and it is the SDK used by the kubernetes provider that needs to be updated to 0.12.1: https://github.com/terraform-providers/terraform-provider-kubernetes/pull/491

**Comment by pdecat** [comment]
I can confirm upgrading to terraform 0.12.1 SDK resolves the issue.

Tested from master with a modified acceptance test to reproduce the issue:

```
diff --git a/kubernetes/resource_kubernetes_resource_quota_test.go b/kubernetes/resource_kubernetes_resource_quota_test.go
index 257db2ea..ebf32f53 100644
--- a/kubernetes/resource_kubernetes_resource_quota_test.go
+++ b/kubernetes/resource_kubernetes_resource_quota_test.go
@@ -249,6 +249,14 @@ func testAccCheckKubernetesResourceQuotaExists(n string, obj *api.ResourceQuota)
 
 func testAccKubernetesResourceQuotaConfig_basic(name string) string {
        return fmt.Sprintf(`
+variable "quota_cpu" {
+  default = 2
+}
+
+variable "quota_memory" {
+  default = "2Gi"
+}
+
 resource "kubernetes_resource_quota" "test" {
   metadata {
     annotations = {
@@ -266,8 +274,8 @@ resource "kubernetes_resource_quota" "test" {
 
   spec {
     hard = {
-      "limits.cpu"    = 2
-      "limits.memory" = "2Gi"
+      "limits.cpu"    = var.quota_cpu
+      "limits.memory" = var.quota_memory
       pods            = 4
     }
   }
```

Running that modified acceptance test with 0.12.0 SDK fails:

```
# make testacc TEST=./kubernetes TESTARGS='-run=TestAccKubernetesResourceQuota_basic -count=1'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./kubernetes -v -run=TestAccKubernetesResourceQuota_basic -count=1 -timeout 120m
=== RUN   TestAccKubernetesResourceQuota_basic
--- FAIL: TestAccKubernetesResourceQuota_basic (0.07s)
    testing.go:568: Step 0 error: config is invalid: 2 problems:
        
        - spec.0.hard.limits.cpu ("74D93920-ED26-11E3-AC10-0800200C9A66"): quantities must match the regular expression '^([+-]?[0-9.]+)([eEinumkKMGTP]*[-+]?[0-9]*)$'
        - spec.0.hard.limits.memory ("74D93920-ED26-11E3-AC10-0800200C9A66"): quantities must match the regular expression '^([+-]?[0-9.]+)([eEinumkKMGTP]*[-+]?[0-9]*)$'
FAIL
FAIL    github.com/terraform-providers/terraform-provider-kubernetes/kubernetes 0.111s
make: *** [GNUmakefile:17: testacc] Error 1
```

Updating to 0.12.1 SDK then re-running the same test passes:

```
# go get github.com/hashicorp/terraform@v0.12.1
# make testacc TEST=./kubernetes TESTARGS='-run=TestAccKubernetesResourceQuota_basic -count=1'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./kubernetes -v -run=TestAccKubernetesResourceQuota_basic -count=1 -timeout 120m
=== RUN   TestAccKubernetesResourceQuota_basic
--- PASS: TestAccKubernetesResourceQuota_basic (4.24s)
PASS
ok      github.com/terraform-providers/terraform-provider-kubernetes/kubernetes 4.277s
```

**Comment by yob** [comment]
Thanks!

**Comment by conet** [comment]
Any plans to release this fix? It would be really nice to have and `v0.12.1` is already released.

**Comment by alexsomesan** [comment]
@conet I merged #491 which fixes this issue. It will be available in the next release.

**Comment by alexsomesan** [comment]
1.8.0 was released, which includes this fix.
Please re-open the issue if problems persist.
